File: javascript/ch1/Ex118.js
// Load with: var ExXXX = require('./ExXXX')
// Run with: ExXXX.xxx()

even = function (x) {
    return x % 2 == 0;
};

mult_invariant = function (b, p, a) {
    if (p == 0) return a;
    if (even(p)) return mult_invariant(b * 2, p / 2, a);
    return mult_invariant(b, p - 1, a + b);
};

times = function (b, p) {
    return mult_invariant(b, p, 0);
};


exports.times = times;

File: haskell/ch1/Ex118.hs
-- :cd C:\Users\Yohan\IdeaProjects\sicp\haskell\ch1
-- :load ExXXX
-- Run with func_name arg1...

module Ex118 where

mult_invariant _ 0 a = a
mult_invariant b p a | even p = mult_invariant (b*2) (div p 2) a
mult_invariant b p a = mult_invariant b (p - 1) (a + b)

times a b = mult_invariant a b 0
